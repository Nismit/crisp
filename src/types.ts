/* eslint-disable @typescript-eslint/no-explicit-any */
type URL = string;
type Template = string;
type Page = number;

export interface Params {
  sort_by?: string;
  types?: string;
  q?: string;
}

/**
 * Accepts an api object and returns whether to keep or remove it from the response
 */
export type FilterFunction = (arg0: any) => boolean;

export interface Request {
  url: URL;
  template: Template;
  page?: Page;
  params?: Params;
}

export interface Headers {
  size: number;
  total: number;
  page: number;
  pages: number;
}

/**
 * An array of the requested api object. Generally based on a template
 */
export type Payload = any[];

/**
 * A callback function that either contains the requested payload or an error. Remember to check if the error is due to cancellation via {@link isCancel}
 * @param {object} args
 * @param {Payload} [args.payload = undefined]
 * @param {Error} [args.error = undefined]
 * @return {undefined}
 * @example
 * collection.get({
 *  number: 48,
 *  callback: function callback(response) {
 *    var payload = response.payload;
 *    var error = response.error;
 *
 *    if (Crisp.isCancel(error)) {
 *      // Can usually ignore
 *      return;
 *    }
 *
 *    if (error) {
 *      // Handle error
 *      return;
 *    }
 *
 *    // Use payload
 *  }
 *});
 */
export type Callback = (arg0: { payload?: Payload; error?: Error }) => void;

export interface Response {
  headers: Headers;
  payload: Payload;
}

export type Cancel = () => void;

/**
 * Defines in what order products are returned
 * @see {@link https://help.shopify.com/themes/liquid/objects/collection#collection-default_sort_by|shopify sort order}
 */
export type CollectionOrder =
  | 'default'
  | 'manual'
  | 'best-selling'
  | 'title-ascending'
  | 'title-descending'
  | 'price-ascending'
  | 'price-descending'
  | 'created-ascending'
  | 'created-descending';

/** */
export type SearchField =
  | 'title'
  | 'handle'
  | 'body'
  | 'vendor'
  | 'product_type'
  | 'tag'
  | 'variant'
  | 'sku'
  | 'author';
