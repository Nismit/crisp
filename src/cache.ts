import { TemplatePayload } from './types';

interface CacheProvider<E = any> {
  /**
   * Retrieve a cached template payload from the cache
   *
   * @param {string} key
   * @returns {(Promise<TemplatePayload<E> | null>)}
   * @memberof CacheProvider
   */
  get(key: string): Promise<TemplatePayload<E> | null>;

  /**
   * Add a template payload to the cache
   *
   * @param {string} key
   * @param {TemplatePayload<E>} template
   * @returns {Promise<void>}
   * @memberof CacheProvider
   */
  set(key: string, template: TemplatePayload<E>): Promise<void>;
}
