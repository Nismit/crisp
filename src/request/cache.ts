import md5 from 'md5';

import { Request, Response } from '../types';

// Cache
let cache = {};

// Takes the params and hashes them
export const hash = (params: Request): string => {
  const json = JSON.stringify(params);
  return `${params.url}-${String(params.page || 1).padStart(3, '0')}-${md5(
    json
  )}`;
};

export const set = async (key: string, response: Response) => {
  cache[key] = response;
};

export const get = async (key: string): Promise<Response | void> => cache[key];

export const clear = async () => {
  cache = {};
};
