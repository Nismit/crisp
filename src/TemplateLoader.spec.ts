import TemplateLoader from './TemplateLoader';
import { CacheProvider } from './cache';
import { TemplatePayload } from './types';

// Cache Mock
type Payload = TemplatePayload<any>;

let store: { [key: string]: Payload } = {};
const get = jest.fn().mockImplementation(key => store[key] || null);
const set = jest.fn().mockImplementation((key, payload) => {
  store[key] = payload;
});

const cache: CacheProvider = {
  get,
  set
};

beforeEach(() => {
  store = {};
  get.mockClear();
  set.mockClear();
});

describe('TemplateLoader()', () => {
  it('exists and returns a function', () => {
    expect(typeof TemplateLoader).toEqual('function');
    expect(typeof TemplateLoader()).toEqual('function');
  });

  it('accepts custom cache and request implementations', async () => {
    const request = jest.fn().mockResolvedValue({
      total: 0,
      pages: 0,
      page: 0,
      size: 0,
      payload: []
    });

    const loadTemplate = TemplateLoader({ cache, request });

    const foo = await loadTemplate('/foo');
    expect(request.mock.calls).toMatchSnapshot('/foo request calls');
    expect(store).toMatchSnapshot('/foo store state');
    expect(get.mock.calls).toMatchSnapshot('/foo cache get calls');
    expect(set.mock.calls).toMatchSnapshot('/foo cache set calls');
    expect(await loadTemplate('/foo')).toEqual(foo);
  });
});
