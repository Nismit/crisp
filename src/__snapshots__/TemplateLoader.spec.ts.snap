// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateLoader() accepts custom cache and request implementations: /foo cache get calls 1`] = `
Array [
  Array [
    "3371046394",
  ],
]
`;

exports[`TemplateLoader() accepts custom cache and request implementations: /foo cache set calls 1`] = `
Array [
  Array [
    "3371046394",
    Object {
      "page": 0,
      "pages": 0,
      "payload": Array [],
      "size": 0,
      "total": 0,
    },
  ],
]
`;

exports[`TemplateLoader() accepts custom cache and request implementations: /foo request calls 1`] = `
Array [
  Array [
    "/foo?",
  ],
]
`;

exports[`TemplateLoader() accepts custom cache and request implementations: /foo store state 1`] = `
Object {
  "3371046394": Object {
    "page": 0,
    "pages": 0,
    "payload": Array [],
    "size": 0,
    "total": 0,
  },
}
`;
